reconcile/cluster_auth_rhidp/integration.py

```diff
@@ -0,0 +1,248 @@
+            [OCMAPIClientConfigurationProtocol, SecretReaderBase], OCMBaseClient
+        ] = init_ocm_base_client,
+    ) -> OcmApis:
+        """Initialize OCM clients for each OCM environment."""
+        return {
+            env.name: init_ocm_base_client(env, self.secret_reader)
+            for env in environments
+        }
+
+    def reconcile(
+        self,
+        dry_run: bool,
+        ocm_apis: OcmApis,
+        current_state: LabelState,
+        desired_state: LabelState,
+    ) -> None:
+        # we iterate via the current state because it refers to the clusters we can act on
+        for label_owner_ref, current_labels in current_state.items():
+            ocm_api = ocm_apis[label_owner_ref.ocm_env]
+            desired_labels = desired_state.get(label_owner_ref, {})
+            if current_labels == desired_labels:
+                continue
+
+            diff_result = diff_mappings(current_labels, desired_labels)
+
+            for label_to_add, value in diff_result.add.items():
+                logging.info([
+                    "create_label",
+                    *label_owner_ref.identity_labels(),
+                    f"{label_to_add}={value}",
+                ])
+                if not dry_run:
+                    add_label(
+                        ocm_api=ocm_api,
+                        label_container_href=label_owner_ref.required_label_container_href(),
+                        label=label_to_add,
+                        value=value,
+                    )
+            for label_to_rm, value in diff_result.delete.items():
+                logging.info([
+                    "delete_label",
+                    *label_owner_ref.identity_labels(),
+                    f"{label_to_rm}={value}",
+                ])
+                if not dry_run:
+                    delete_label(
+                        ocm_api=ocm_api,
+                        label_container_href=label_owner_ref.required_label_container_href(),
+                        label=label_to_rm,
+                    )
+            for label_to_update, diff_pair in diff_result.change.items():
+                value = diff_pair.desired
+                logging.info([
+                    "update_label",
+                    *label_owner_ref.identity_labels(),
+                    f"{label_to_update}={value}",
+                ])
+                if not dry_run:
+                    update_label(
+                        ocm_api=ocm_api,
+                        label_container_href=label_owner_ref.required_label_container_href(),
+                        label=label_to_update,
+                        value=value,
+                    )
+
+    def run(self, dry_run: bool) -> None:
+        """Run the integration."""
+        gql_api = gql.get_api()
+        clusters = self.get_clusters(gql_api.query)
+        environments = self.get_environments(gql_api.query)
+        ocm_apis = self.init_ocm_apis(environments, init_ocm_base_client)
```

Comment:
`ocm_apis` require close session after use, since we are not using multi threads, we can also create a single session in `run` and pass it to all `ocm_apis`, and wrap it using `with` or use `defer` to close.

Reply:
Fixed for this integration and `ocm-labels` in [2845fa0](https://github.com/app-sre/qontract-reconcile/pull/4338/commits/2845fa0232c041a58b8c2b93874a14cf6e14243e)